akka-actor/src/main/java/akka/japi/pf/Reflect.java

```diff
@@ -0,0 +1,134 @@
+/**
+ * Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.japi.pf;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.Objects;
+
+import sun.reflect.ConstantPool;
+
+/**
+ * INTERNAL API
+ */
+public class Reflect {
+    private static final Method getConstantPoolMethod;
+
+    static {
+        try {
+            getConstantPoolMethod = Class.class.getDeclaredMethod("getConstantPool");
+            getConstantPoolMethod.setAccessible(true);
+        } catch (NoSuchMethodException e) {
+            throw new RuntimeException("Couldn't find the 'getConstantPool' method - are you using the SUN JVM?",
+                                       e);
+        }
+    }
+
+    /**
+     * Resolve the generic argument types of a lamb instance (even though types are not reified in Java)<br/>
+     * Example: If a class/method is handed a Function&lt;Foo, Bar> then resolveGenericLambdaArgumentType for argument 0 (the only argument) will return String.class:
+     * <pre>
+     * Function&lt;String, Boolean> myStrToBoolFunc = strArg -> true;
+     * assertThat(Reflect.resolveGenericLambdaArgumentType(myStrToBoolFunc, 0), equalTo(String.class));
+     * </pre>
+     * Example of a lambda with multiple arguments:
+     * <pre>
+     * BiConsumer&lt;Integer, Long> biConsumer = (i, l) -> {};
+     * assertThat(Reflect.resolveGenericLambdaArgumentType(biConsumer, 0), equalTo(Integer.class));
+     * assertThat(Reflect.resolveGenericLambdaArgumentType(biConsumer, 1), equalTo(Long.class));
+     * </pre>
+     */
+    public static Class<?> resolveLambdaArgumentType(Object object, int argumentIndex) {
+        Objects.requireNonNull(object,
+                               "object");
+        if (argumentIndex < 0) {
+            throw new IllegalArgumentException("ArgumentIndex must be >= 0");
+        }
+
+        Type genericInterface = object.getClass()
+                                      .getGenericInterfaces()[0];
+        if (genericInterface instanceof ParameterizedType) {
+            Type type = ((ParameterizedType) genericInterface).getActualTypeArguments()[argumentIndex];
+            if (type instanceof ParameterizedType) {
+                return (Class<?>) ((ParameterizedType)type).getRawType();
+            } else {
+                return (Class<?>) type;                
+            }
+        }
+        else {
+            String[] methodRef = resolveMethodRef(object, "Failed to resolve generic type arguments for " + object + " of type " + genericInterface.toString());
+            try {
+                if (methodRef[1].startsWith("lambda")) {
+                    // This is an inline lambda expression, e.g. (String s) -> s.isEmpty()
+                    String argumentType = jdk.internal.org.objectweb.asm.Type.getArgumentTypes(methodRef[2])[argumentIndex].getClassName();
```

Comment:
It seems compiled type information for lambdas isn't available runtime otherwise.

The only alternative I see is providing a `.matchUnsafe(Class<A>, Function<B,C>)` that allows one to match one class to an apply of a totally different type. That's what's needed unfortunately when dealing with generic types in Java, and why the unrelated types were originally introduced in https://github.com/akka/akka/issues/16245 . It seems to have been overlooked back then that by saying `<P> Type1<? extends P>, Type2<? extends P>`, the compiler will happily drop `P` all the way down to `Object`.

So, if using internal JDK is ill-advised, it's choosing between
1. Having no compile safety at all on these methods (as it is currently)
2. Have two methods, one which is compile-safe (but can't be used with generic types), and another which isn't compile-safe and does blind casting underneath (as it is currently)

I'd vote for (2), but we'd have to find some not-too-frightening names.